### ISSUE TITLE
get_url does not set mode if file already exists

### ISSUE THREAD

**Issue by feliksik** [ansible/ansible#11821]
```
- name: fetch url
  get_url: url=https://example.com/file.txt dest=/file.txt mode=0755
```

Behavior: if /file.txt already exists on the host, the mode is _not_ set to 0755. 

Expected behavior: mode ends up being 0755, no matter what the mode was before running the task.

**Comment by muffl0n** [comment]
`get_url` does not support `mode` at all. Though it says so in the examples:

```
- name: download foo.conf
  get_url: url=http://example.com/path/file.conf dest=/etc/foo.conf mode=0440
```

It seems like the example exists since the [initial commit](https://github.com/ansible/ansible/commit/b4d481107730956200d3cccf784a9c98fe1dc6f2#diff-b4657d275e143d853912c603e58eb806R34) but has never been implemented.

Edit: Forget that. :( Just learned about `module.set_fs_attributes_if_different` and that it should do exactly what you described. :)

**Comment by muffl0n** [comment]
@feliksik: Which version are you using? With current devel (c3ce140dd2) it does work as expected:

```
svs@kuzco ~/ansible_testing/get_url $ ansible -i hosts localhost -m get_url -a "url=http://www.example.com dest=/tmp"
localhost | SUCCESS => {
    "changed": true, 
    "checksum": "0e973b59f476007fd10f87f347c3956065516fc0", 
    "dest": "/tmp/index.html", 
    "gid": 1000, 
    "group": "svs", 
    "md5sum": "09b9c392dc1f6e914cea287cb6be34b0", 
    "mode": "0644", 
    "msg": "OK (1270 bytes)", 
    "owner": "svs", 
    "sha256sum": "", 
    "size": 1270, 
    "src": "/tmp/tmpU376v0", 
    "state": "file", 
    "uid": 1000, 
    "url": "http://www.example.com"
}
svs@kuzco ~/ansible_testing/get_url $ l /tmp/index.html 
-rw-r--r-- 1 svs svs 1270  4. Aug 10:29 /tmp/index.html
svs@kuzco ~/ansible_testing/get_url $ ansible -i hosts localhost -m get_url -a "url=http://www.example.com dest=/tmp mode=0775"
localhost | SUCCESS => {
    "changed": true, 
    "checksum": "0e973b59f476007fd10f87f347c3956065516fc0", 
    "dest": "/tmp/index.html", 
    "gid": 1000, 
    "group": "svs", 
    "md5sum": "09b9c392dc1f6e914cea287cb6be34b0", 
    "mode": "0775", 
    "msg": "OK (1270 bytes)", 
    "owner": "svs", 
    "sha256sum": "", 
    "size": 1270, 
    "src": "/tmp/tmpxkcTAc", 
    "state": "file", 
    "uid": 1000, 
    "url": "http://www.example.com"
}
svs@kuzco ~/ansible_testing/get_url $ l /tmp/index.html 
-rwxrwxr-x 1 svs svs 1270  4. Aug 10:29 /tmp/index.html
```

**Comment by feliksik** [comment]
I am using 1.9.2 -- can't test with newest version now :-s

**Comment by muffl0n** [comment]
I just checked out 1.9.2 and could not reproduce the problem either:

```
svs@kuzco ~/ansible_testing/get_url $ ansible --version
ansible 1.9.2 (detached HEAD f974ff6972) last updated 2015/08/04 13:08:28 (GMT +200)
  lib/ansible/modules/core: (detached HEAD ea913bada2) last updated 2015/08/04 13:09:32 (GMT +200)
  lib/ansible/modules/extras: (detached HEAD f291e9fcc3) last updated 2015/08/04 13:09:44 (GMT +200)
  v2/ansible/modules/core: (detached HEAD 85c8a892c8) last updated 2015/08/04 13:09:59 (GMT +200)
  v2/ansible/modules/extras: (detached HEAD 70ea058563) last updated 2015/08/04 13:10:11 (GMT +200)
  configured module search path = None
svs@kuzco ~/ansible_testing/get_url $ ansible -i hosts localhost -m get_url -a "url=http://www.example.com dest=/tmp"
localhost | success >> {
    "changed": true,
    "checksum": "0e973b59f476007fd10f87f347c3956065516fc0",
    "dest": "/tmp/index.html",
    "gid": 1000,
    "group": "svs",
    "md5sum": "09b9c392dc1f6e914cea287cb6be34b0",
    "mode": "0644",
    "msg": "OK (1270 bytes)",
    "owner": "svs",
    "sha256sum": "",
    "size": 1270,
    "src": "/tmp/tmpMBxKsJ",
    "state": "file",
    "uid": 1000,
    "url": "http://www.example.com"
}

svs@kuzco ~/ansible_testing/get_url $ l /tmp/index.html 
-rw-r--r-- 1 svs svs 1270  4. Aug 13:11 /tmp/index.html
svs@kuzco ~/ansible_testing/get_url $ ansible -i hosts localhost -m get_url -a "url=http://www.example.com dest=/tmp mode=0755"
localhost | success >> {
    "changed": true,
    "checksum": "0e973b59f476007fd10f87f347c3956065516fc0",
    "dest": "/tmp/index.html",
    "gid": 1000,
    "group": "svs",
    "md5sum": "09b9c392dc1f6e914cea287cb6be34b0",
    "mode": "0755",
    "msg": "OK (1270 bytes)",
    "owner": "svs",
    "sha256sum": "",
    "size": 1270,
    "src": "/tmp/tmpaV4dhC",
    "state": "file",
    "uid": 1000,
    "url": "http://www.example.com"
}

svs@kuzco ~/ansible_testing/get_url $ l /tmp/index.html 
-rwxr-xr-x 1 svs svs 1270  4. Aug 13:11 /tmp/index.html
```

If you have the time to test with the latest version just check this page: http://docs.ansible.com/ansible/intro_installation.html#running-from-source

**Comment by sysadmin75** [comment]
I can't reproduce either using latest devel and latest 1.9.3 version.

Management Node: Centos 6
Managed Node: Centos 6

**Comment by muffl0n** [comment]
Voting to close this issue. Cannot reproduce and OP doesn't react.

**Comment by feliksik** [comment]
Sorry for not responding earlier, @muffl0n . You are apparently testing with dest=/tmp . This works for me (ansible 1.9.2). 

However, try with dest=/tmp/index.html and it will fail (ansible 1.9.2). 

To save you some time, here a playbook that fails on the last step to reproduce the bug: 

```
- name: test https://github.com/ansible/ansible/issues/11821
  hosts: all
  vars:
    url: http://www.example.com/index.html
#    file: /tmp # this works fine
    file: /tmp/index.html # this fails
  tasks:
  - name: get a file with default (644) permissions
    get_url: url={{url}} dest={{file}}
  - name: get a file, now with explicit permissions
    get_url: url={{url}} dest={{file}} mode=0755
  - name: stat
    stat: path={{file}}
    register: p
  - name: check if file is now executable
    shell: "true"
    failed_when: p.stat.xusr == false
```

Note how the second get_url reports success and mentions mode 0664 (although the mode=0755 parameter is given in the task)

```
ok: [test] => {"changed": false, "dest": "/tmp/index.html", "gid": 1000, "group": "client", "mode": "0664", "msg": "file already exists", "owner": "client", "size": 1270, "state": "file", "uid": 1000, "url": "http://www.example.com/index.html"}
```

**Comment by muffl0n** [comment]
You are right, when specifying a file as dest it breaks as described:

```
$ ansible -i hosts localhost -m get_url -a "url=http://www.example.com dest=/tmp/index.html "
localhost | success >> {
    "changed": true, 
    "checksum": "0e973b59f476007fd10f87f347c3956065516fc0", 
    "dest": "/tmp/index.html", 
    "gid": 1000, 
    "group": "muffl0n", 
    "md5sum": "09b9c392dc1f6e914cea287cb6be34b0", 
    "mode": "0664", 
    "msg": "OK (1270 bytes)", 
    "owner": "muffl0n", 
    "sha256sum": "", 
    "size": 1270, 
    "src": "/tmp/tmpRA5jO3", 
    "state": "file", 
    "uid": 1000, 
    "url": "http://www.example.com"
}
```

```
$ ansible -i hosts localhost -m get_url -a "url=http://www.example.com dest=/tmp/index.html mode=0755"
localhost | success >> {
    "changed": false, 
    "dest": "/tmp/index.html", 
    "gid": 1000, 
    "group": "muffl0n", 
    "mode": "0664", 
    "msg": "file already exists", 
    "owner": "muffl0n", 
    "size": 1270, 
    "state": "file", 
    "uid": 1000, 
    "url": "http://www.example.com"
}
```

It's that way because get_url does a fail fast if the file already exists. See https://github.com/ansible/ansible-modules-core/blob/devel/network/basics/get_url.py#L279

It does redownload the file and set it's mode if you set force=true:

```
$ ansible -i hosts localhost -m get_url -a "url=http://www.example.com dest=/tmp/index.html mode=0755 force=true"
localhost | success >> {
    "changed": true, 
    "checksum": "0e973b59f476007fd10f87f347c3956065516fc0", 
    "dest": "/tmp/index.html", 
    "gid": 1000, 
    "group": "muffl0n", 
    "md5sum": "09b9c392dc1f6e914cea287cb6be34b0", 
    "mode": "0755", 
    "msg": "OK (1270 bytes)", 
    "owner": "muffl0n", 
    "sha256sum": "", 
    "size": 1270, 
    "src": "/tmp/tmpc4dM65", 
    "state": "file", 
    "uid": 1000, 
    "url": "http://www.example.com"
}
```

I don't feel that this is a bug because get_urls primary task is not to adjust file permissions but to just download the file.

Either set `force=true` to redownload the file or just set the file mode explicitly after calling `get_url`. 
E.g.:

```
$ ansible -i hosts localhost -m file -a "dest=/tmp/index.html mode=0777"
localhost | success >> {
    "changed": true, 
    "gid": 1000, 
    "group": "muffl0n", 
    "mode": "0777", 
    "owner": "muffl0n", 
    "path": "/tmp/index.html", 
    "size": 1270, 
    "state": "file", 
    "uid": 1000
}
```

**Comment by muffl0n** [comment]
Btw: Behaviour is the same in 1.9.2 and devel.

**Comment by feliksik** [comment]
@muffl0n 

> I don't feel that this is a bug because get_urls primary task is not to adjust file permissions but to just download the file.

I kindly disagree; having force=false by default is a sensible idea, but the mode is just a declaration of the desired end-state, which is _the purpose of CM_. If there is such a thing as a primary vs. secondary purpose, then let's document 'secondary purpose' functionality which "may or may not work" ;-) 

It would, by the way, help if the mode directive was documented in the table of http://docs.ansible.com/ansible/get_url_module.html , not only in the example. 

In particular, having the semantics different for dest=/tmp and dest=/tmp/file.txt is an implementation quirk, thus IMHO a bug and not a feature.

**Comment by muffl0n** [comment]
I totally agree that the behavior should not differ between specifying a dir or a file as dest.

**Comment by rewritten** [comment]
This is also a problem when using `file` option as per the documentation at http://docs.ansible.com/ansible/get_url_module.html:

> all arguments accepted by the file module also work here

If a task has:

``` yaml
get_url:
  url: somewhere
  dest: some file
  checksum: "..."
  mode: ug+x,u+s
  group: wheel
```

and the file part fails because the current user does not have right to assign the wheel group, the downloaded file is left there, and on next execution the precondition (dest+checksum) succeeds so the whole task is not executed.

IMHO this needs the tag `pending_closure` to be removed.

**Comment by rewritten** [comment]
Either that or remove the `file` abilities of `get_url`.

**Comment by abadger** [comment]
Fixed and changes pushed to devel, stable-2.0 and stable-1.9 branches

**Comment by abadger** [comment]
Here's the commit on devel (of the ansible-modules-core repo: https://github.com/ansible/ansible-modules-core/commit/d71b9bae893e5cb860eb7717af93d00363b3890c )
